Aklite and composectl sotactl v95

mike-sul opened the pull request:
# Aktualizr-lite

## New Features

1. Added support for EFI boot firmware updates, including functionality to get/set bootloader control environment variables using the `fioefi_setenv` and `fioefi_printenv` utilities.

2. Added new CLI client commands: `check`, `pull`, `install`, and `run`. These individual commands allow users to pull, install, and apply updates from different contexts.

3. Introduced the `--src-dir` parameter for the `aktualizr-lite` CLI client, enabling offline updates.

4. Automatic downgrade prevention.

5. Added support for user-initiated rollbacks [experimental]

## Improvements

1. Integrated `composectl` for all operations involving Compose Apps, such as pruning, checking if running, etc.

2. Optimized app management to minimize the number of actions performed on Compose Apps during the update cycle.

3. Made app stopping before an update configurable.

4. Introduced several changes to the API (refer to the commit message for details).

6. Moved the custom SOTA client example from the `aktualizr-lite` repository to a dedicated repository: [`https://github.com/foundriesio/sotactl`](https://github.com/foundriesio/sotactl).

7. Transitioned the `aklite` daemon to API-based usage.

## Testing

1. Added a Compose-based environment for developing and testing `aktualizr-lite` end-to-end in a containerized setup against a real factory.


# Composectl

## New Features

1. Added the `publish` command to package a Compose App as a container image and upload it to a container registry.

2. Added the `show` command to print app manifest and Compose project details.

3. Introduced App bundle indexing, which includes hash generation for each item in the bundle. The hashes are verified during app OTA updates, installation, and starting. Additionally, the maximum app bundle size is now limited to 1GB.

## Bug Fixes

- Resolved several minor issues, mainly related to app blob pruning and checking whether an app is installed and running.  
  See the commit message for more details.

## Testing

1. Added a Compose-based environment for development and end-to-end testing of `composectl` in a containerized setup.

2. Implemented a series of end-to-end tests, including tests for edge cases.

quaresmajose commented:
@mike-sul can you please rebase

mike-sul commented:
@caiotpereira @vanmaegima FYI.

This the candidate for v95 (aklite + composectl).

I added the release notes into the component directory. The same notes are added as this PR description.

https://github.com/foundriesio/meta-lmp/blob/aklite_and_composectl_sotactl_v95/meta-lmp-base/recipes-sota/aktualizr/release-notes.md

https://github.com/foundriesio/meta-lmp/blob/aklite_and_composectl_sotactl_v95/meta-lmp-base/recipes-containers/composeapp/release-notes.md

ricardosalveti commented:
Can we instead have the release notes on the respective project tags and a link pointing to them as part of the commit message? Better to handle the individual component release notes in the component repository itself.

ricardosalveti commented:
It is also currently in draft mode.

mike-sul commented:
> It is also currently in draft mode.

It is WIP, will update the PR mode once it is ready.

mike-sul commented:
> Can we instead have the release notes on the respective project tags and a link pointing to them as part of the commit message?

I guess we can. The problem is that we actually don't have a notion of release for aklite because it is not "self-contained" or "self-sufficient" - it works solely in a scope of LmP and is part of LmP releases.
However, I think we can simply have notes on a tag without doing github releases.

ricardosalveti commented:
> However, I think we can simply have notes on a tag without doing github releases.

Yeah, that would work better, as we don't usually keep release notes as part of this repo.

mike-sul commented:
> Can we instead have the release notes on the respective project tags and a link pointing to them as part of the commit message? Better to handle the individual component release notes in the component repository itself.

@ricardosalveti Unfortunately, a big comment (release notes) on a repo/project tag is not rendered nicely by github. It looks like github designed to render release notes only for a repo/project releases in github. We don't do aklite not composectl releases in github.

![image](https://github.com/user-attachments/assets/3a36d612-7f12-47e6-bef9-9a62e39b4aa6)

![image](https://github.com/user-attachments/assets/c9d07365-ac5a-4b0e-a0a6-7fdba1dbac3f)

detsch commented:
@mike-sul 
`4. Added support for user-initiated rollbacks.` -> this is not included in `v95`

What's missing:

_- Automatic downgrade prevention: a downgrade operation is not performed anymore when all targets listed in TUF metadata have a version lower than the current running target. This change can be reverted by enabling the `auto-downgrade` package config in `aktualizr` recipe._


In line with the above comment, perhaps we can add  https://github.com/foundriesio/meta-lmp/commit/8cd6405715973b90873fbc5d792fa2ae52a25633 to the current PR.

mike-sul commented:
> @mike-sul `4. Added support for user-initiated rollbacks.` -> this is not included in `v95`
> 
> What's missing:
> 
> _- Automatic downgrade prevention: a downgrade operation is not performed anymore when all targets listed in TUF metadata have a version lower than the current running target. This change can be reverted by enabling the `auto-downgrade` package config in `aktualizr` recipe._
> 
> In line with the above comment, perhaps we can add [8cd6405](https://github.com/foundriesio/meta-lmp/commit/8cd6405715973b90873fbc5d792fa2ae52a25633) to the current PR.

Fixed.

detsch commented:
@mike-sul 

This can be re-added, as we decided to add user initiated rollbacks. Perhaps highlighting that is an experimental feature.

> 4. Added support for user-initiated rollbacks

What about adding https://github.com/foundriesio/meta-lmp/commit/8cd6405715973b90873fbc5d792fa2ae52a25633?

mike-sul commented:
> This can be re-added, as we decided to add user initiated rollbacks. Perhaps highlighting that is an experimental feature.

It is included in the v95 branch of aklite as well as mentioned in the commit message. I will just add it to the PR comment/description.

> What about adding [8cd6405](https://github.com/foundriesio/meta-lmp/commit/8cd6405715973b90873fbc5d792fa2ae52a25633)?

We can either append this commit to this PR or just create another PR to meta-lmp, it is up to you.

detsch commented:
>> What about adding [8cd6405](https://github.com/foundriesio/meta-lmp/commit/8cd6405715973b90873fbc5d792fa2ae52a25633)?

> We can either append this commit to this PR or just create another PR to meta-lmp, it is up to you.

I've added it to the current PR.
